File: backend/webdav/src/main/java/com/mail/ann/backend/webdav/CommandMoveOrCopyMessages.java
package com.mail.ann.backend.webdav;


import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

import com.mail.ann.mail.MessagingException;
import com.mail.ann.mail.store.webdav.WebDavFolder;
import com.mail.ann.mail.store.webdav.WebDavMessage;
import com.mail.ann.mail.store.webdav.WebDavStore;
import org.jetbrains.annotations.NotNull;
import timber.log.Timber;


class CommandMoveOrCopyMessages {
    private final WebDavStore webDavStore;


    CommandMoveOrCopyMessages(WebDavStore webDavStore) {
        this.webDavStore = webDavStore;
    }

    Map<String, String> moveMessages(@NotNull String sourceFolderServerId, @NotNull String targetFolderServerId,
            @NotNull List<String> messageServerIds) throws MessagingException {
        return moveOrCopyMessages(sourceFolderServerId, targetFolderServerId, messageServerIds, false);
    }

    Map<String, String> copyMessages(@NotNull String sourceFolderServerId, @NotNull String targetFolderServerId,
            @NotNull List<String> messageServerIds) throws MessagingException {
        return moveOrCopyMessages(sourceFolderServerId, targetFolderServerId, messageServerIds, true);
    }

    private Map<String, String> moveOrCopyMessages(String srcFolder, String destFolder, Collection<String> uids,
            boolean isCopy) throws MessagingException {
        WebDavFolder remoteSrcFolder = null;
        WebDavFolder remoteDestFolder = null;

        try {
            remoteSrcFolder = webDavStore.getFolder(srcFolder);

            List<WebDavMessage> messages = new ArrayList<>();

            for (String uid : uids) {
                messages.add(remoteSrcFolder.getMessage(uid));
            }

            if (messages.isEmpty()) {
                Timber.i("processingPendingMoveOrCopy: no remote messages to move, skipping");
                return null;
            }

            remoteSrcFolder.open();

            Timber.d("processingPendingMoveOrCopy: source folder = %s, %d messages, " +
                    "destination folder = %s, isCopy = %s", srcFolder, messages.size(), destFolder, isCopy);


            remoteDestFolder = webDavStore.getFolder(destFolder);

            if (isCopy) {
                return remoteSrcFolder.copyMessages(messages, remoteDestFolder);
            } else {
                return remoteSrcFolder.moveMessages(messages, remoteDestFolder);
            }
        } finally {
            closeFolder(remoteSrcFolder);
            closeFolder(remoteDestFolder);
        }
    }

    private static void closeFolder(WebDavFolder folder) {
        if (folder != null) {
            folder.close();
        }
    }
}

File: backend/imap/src/main/java/com/mail/ann/backend/imap/CommandMoveOrCopyMessages.kt
package com.mail.ann.backend.imap

import com.mail.ann.mail.MessagingException
import com.mail.ann.mail.store.imap.ImapFolder
import com.mail.ann.mail.store.imap.ImapStore
import com.mail.ann.mail.store.imap.OpenMode
import timber.log.Timber

internal class CommandMoveOrCopyMessages(private val imapStore: ImapStore) {

    fun moveMessages(
        sourceFolderServerId: String,
        targetFolderServerId: String,
        messageServerIds: List<String>
    ): Map<String, String>? {
        return moveOrCopyMessages(sourceFolderServerId, targetFolderServerId, messageServerIds, false)
    }

    fun copyMessages(
        sourceFolderServerId: String,
        targetFolderServerId: String,
        messageServerIds: List<String>
    ): Map<String, String>? {
        return moveOrCopyMessages(sourceFolderServerId, targetFolderServerId, messageServerIds, true)
    }

    private fun moveOrCopyMessages(
        srcFolder: String,
        destFolder: String,
        uids: Collection<String>,
        isCopy: Boolean
    ): Map<String, String>? {
        var remoteSrcFolder: ImapFolder? = null
        var remoteDestFolder: ImapFolder? = null

        return try {
            remoteSrcFolder = imapStore.getFolder(srcFolder)

            if (uids.isEmpty()) {
                Timber.i("moveOrCopyMessages: no remote messages to move, skipping")
                return null
            }

            if (!remoteSrcFolder.exists()) {
                throw MessagingException("moveOrCopyMessages: remoteFolder $srcFolder does not exist", true)
            }

            remoteSrcFolder.open(OpenMode.READ_WRITE)
            if (remoteSrcFolder.mode != OpenMode.READ_WRITE) {
                throw MessagingException(
                    "moveOrCopyMessages: could not open remoteSrcFolder $srcFolder read/write", true
                )
            }

            val messages = uids.map { uid -> remoteSrcFolder.getMessage(uid) }

            Timber.d(
                "moveOrCopyMessages: source folder = %s, %d messages, destination folder = %s, isCopy = %s",
                srcFolder, messages.size, destFolder, isCopy
            )

            remoteDestFolder = imapStore.getFolder(destFolder)
            if (isCopy) {
                remoteSrcFolder.copyMessages(messages, remoteDestFolder)
            } else {
                remoteSrcFolder.moveMessages(messages, remoteDestFolder)
            }
        } finally {
            remoteSrcFolder?.close()
            remoteDestFolder?.close()
        }
    }
}
